instrumentation/micrometer/micrometer-1.5/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/micrometer/v1_5/AsyncInstrumentRegistry.java

```diff
@@ -36,83 +44,230 @@
       @Nullable T obj,
       ToDoubleFunction<T> objMetric) {
 
+    String name = meterId.getName();
     synchronized (gauges) {
-      GaugeMeasurementsRecorder recorder =
+      // use the gauges map as lock for the recorder state - this way all gauge-related mutable
+      // state will always be accessed in synchronized(gauges)
```

Comment:
it's not clear to me if it's important to use the same lock for the outer and inner HashMaps? (relatedly wondering if the outer and inner maps could be ConcurrentHashMaps for simplicity, I think these methods aren't called heavily but not sure?)

Reply:
In `removeMeasurement()`, there is a short time window between removing a single measurement and the whole recorder where potentially a new measurement could be added. I opted for correctness in this case, exactly because these methods are not heavily called - and we can easily afford covering the whole method with a `synchronized` statement on the same lock.

Reply:
> there is a short time window between removing a single measurement and the whole recorder where potentially a new measurement could be added

can you add this as a comment, maybe above `@GuardedBy("gauges")`?

Reply:
Added in https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/5017/commits/821a1226c505c6eaf21ce4e08fda52901d143950